mcs/class/corlib/System.Reflection/MonoMethod.cs

```diff
@@ -296,6 +296,15 @@ internal override int GetParametersCount ()
 		[DebuggerStepThrough]
 		public override Object Invoke (Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
 		{
+			if (!IsStatic) {
+				if (!DeclaringType.IsInstanceOfType (obj)) {
```

Comment:
I'd rewrite these conditions to 
```csharp
if (!IsStatic) {
    if (obj == null) 
        throw new TargetException ("Non-static method requires a target.");
    if (!DeclaringType.IsInstanceOfType (obj))
        throw new TargetException ("Object does not match target type.");
}
```
so !IsStatic and obj == null won't need to call that icall `IsInstanceOfType `

Reply:
That's a pointless optimization as the success path will have the same amount of checks.